main.go

```diff
@@ -0,0 +1,186 @@
+package main
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"strings"
+
+	"gopkg.in/yaml.v2"
+)
+
+type ConfigSettings struct {
+	Port  string   `yaml:"port"`
+	Repos []string `yaml:"repos"`
+}
+
+// Global Configuration for the service
+var Config ConfigSettings
+
+func main() {
+	var err error
+	configFile := "config.yaml"
+	if len(os.Args) == 1 {
+		fmt.Println("no configuration file specified, defaulting to config.yaml")
+	}
+	if len(os.Args) > 1 {
+		configFile = os.Args[1]
+	}
+
+	// Get the config file on startup to prevent repeated IO/ parsing
+	yamlData, err := ioutil.ReadFile(configFile)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	Config = ConfigSettings{}
+
+	err = yaml.Unmarshal(yamlData, &Config)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	if len(Config.Port) == 0 {
+		fmt.Println("no port specified in the yaml config file, defaulting to 8000")
+		Config.Port = "8000"
+	}
+
+	if len(Config.Repos) == 0 {
+		log.Fatal(errors.New("no git repos were configured in the yaml config file, at least one repo must be listed for the service to query"))
+	}
+
+	http.HandleFunc("/search", func(w http.ResponseWriter, r *http.Request) {
+		var err error
+
+		if r.Method != "GET" {
+			errorResp(w, fmt.Sprintf("incorect request method %s only the GET method is allowed", r.Method))
+			return
+		}
+
+		req := &SearchRequest{}
+		req.SearchTerm = r.URL.Query().Get("q")
+		if len(req.SearchTerm) == 0 {
+			errorResp(w, "no search term was found, query string must have a 'q' parameter which must be at least 1 character long")
+			return
+		}
+		req.User = r.URL.Query().Get("user")
+
+		searchResp, err := search(req)
+		if err != nil {
+			errorResp(w, fmt.Sprintf("search query could not be compleated, %s", err))
+			return
+		}
+
+		respJSON, err := json.Marshal(searchResp)
+		if err != nil {
+			errorResp(w, fmt.Sprintf("could not marshal query response %s", err))
+			return
+		}
+
+		w.Header().Add("Content-Type", "application/json")
+		w.Write(respJSON)
+	})
+
+	fmt.Println("starting service on port :" + Config.Port)
+	log.Fatal(http.ListenAndServe(":"+Config.Port, nil))
+}
+
+type SearchRequest struct {
+	SearchTerm string `json:"SearchTerm"`
+	User       string `json:"User"`
+}
+
+type Result struct {
+	FileURL string `json:"fileURL"`
+	Repo    string `json:"repo"`
+}
+
+type SearchResponse struct {
+	Results []Result
+}
+
+func search(req *SearchRequest) (*SearchResponse, error) {
+
+	url := "https://api.github.com/search/code?q=" + req.SearchTerm + "+"
+
+	// prevent lots of heap reallocations
+	repos := make([]string, 0, len(Config.Repos))
+	for _, repo := range Config.Repos {
+		// if a user was specified filter only by that specific user
+		if repo[:len(req.User)] == req.User && repo[len(req.User)] == '/' {
```

Comment:
You should replace both of these conditions with one `strings.HasPrefix`

Reply:
The first condition ```repo[:len(req.User)] == req.User``` can indeed be simply replaced with ```strings.HasPrefix```. The second condition ```repo[len(req.User)] == '/'``` is necessary to prevent bugs caused by filtering with user names which are prefixes of other names (e.g. bja & bjatkin). This section of code has been slightly refactored to reflect these needs as well as use the more standard ```strings.HasPrefix``` method for checking string prefixes.

Reply:
Not quite, You can indeed replace them both one one like so:

```suggestion
		if strings.HasPrefix(repo, req.User+"/"){
```